### DISCUSSION TITLE
Trying to separate Statamic into "content" and "structure" directories.

### DISCUSSION THREAD

**Discussion by Swennet** [statamic/cms#8344] [Troubleshooting] [answered]
Hi, I'm attempting to fully separate "content" and "structure" for Statamic into separate directories. The end goal being to store "variable" stuff like content in a gitignored directory, and move all structure-related files/settings into one place.

My directory tree looks like this:

- app
- ...
- content
  - collections
  - forms
  - globals
  - navigation
  - taxonomies
  - trees 
- resources
  - blueprints --> Needs to be moved to ./statamic_structure
  - css
  - images
  - users --> Needs to be moved to ./statamic_structure
  - views
  - preferences.yaml --> Need to be moved to ./statamic_structure
- statamic_structure
  - assets
  - collections
  - taxonomies
  - trees
- ...

Is there any way to configure were these directories/files inside `./resources` are stored? Also, if this is a really bad idea, then please also explain why 😄

**Comment by jasonvarga** [comment]
You could alternatively adjust your gitignore file to ignore the directories you want ignored.

**Comment by ryanmitchell** [comment, marked as answer]
For blueprints/fieldsets you can change the directory in your app service provider:

```
        $this->app->bind(\Statamic\Fields\BlueprintRepository::class, function () {
            return (new \Statamic\Fields\BlueprintRepository)
                ->setDirectory(resource_path('blueprints'))
                ->setFallback('default', function () {
                    return \Statamic\Facades\Blueprint::makeFromFields([
                        'content' => ['type' => 'markdown', 'localizable' => true],
                    ]);
                });
        });

        $this->app->singleton(\Statamic\Fields\FieldsetRepository::class, function () {
            return (new \Statamic\Fields\FieldsetRepository)
                ->setDirectory(resource_path('fieldsets'));
        });
```

For preferences you'd need to create your own class extending `Statamic\Preferences\DefaultPreferences` and bind it in your app service provider - changing the constructor method should be sufficient.

**Reply by ryanmitchell** [reply]
Sorry I missed users - you can change the directory in config/statamic/stache.php - the key is stores.users

**Comment by Swennet** [comment]
@ryanmitchell Thanks! That worked for Blueprints and Fieldsets.

However, the stores.users configuration seems to be only for user accounts, not for groups and roles. Am I missing something, or is that configured somewhere else?

I have two "users" directories:
  - /users -> contains accounts
  - /resources/users -> contains groups and roles

--

For the preferences.yaml file I've extended the DefaultPreferences class with a new constructor method:

```
<?php

namespace App\Statamic;

use Statamic\Facades\YAML;

class CustomDefaultPreferences extends \Statamic\Preferences\DefaultPreferences
{
    /**
     * Instantiate default preferences instance.
     */
    public function __construct()
    {
        $this->path = base_path('statamic_structure/preferences.yaml');

        $this->preferences = YAML::file($this->path)->parse();
    }
}
```

Although I'm unsure how to bind it in my AppServiceProvider. For now I tried this, which doesn't seem to work.

```
<?php

namespace App\Providers;

use App\Statamic\CustomDefaultPreferences;
use Illuminate\Support\ServiceProvider;

class AppServiceProvider extends ServiceProvider
{
    /**
     * Register any application services.
     */
    public function register(): void
    {
        $this->app->bind(CustomDefaultPreferences::class);
    }
}
```

**Reply by ryanmitchell** [reply]
For the preferences you bind as follows: `$this->app->bind(\Statamic\Preferences\DefaultPreferences::class, CustomDefaultPreferences::class);`

The groups and roles paths are defined in config/statamic/users - see the repositories.files.users keys

**Comment by Swennet** [comment]
Can't believe I overlooked the groups and roles. Thanks for the help!
